### ISSUE TITLE
Enum with custom toString() breaks tests

### ISSUE THREAD

**Issue by sormuras** [square/javapoet#343]
If you implement custom `toString()` in `AnnotationSpecTest` like
```java
  public enum Breakfast {
    WAFFLES, PANCAKES;
    public String toString() { return "Lunch#" + name(); }
  }
```
some tests fail due to the assumption of `toString()` returning always `name()`.

I can fix the test code by using `name()` explictely or introduce an `$E` token for enum references. May even take the #317 issue into account.

What do you think?

**Comment by sormuras** [comment]
`format("$E", Breakfast.WAFFLES)` could be a shortcut for
`format("$T.$L", Breakfast.class, Breakfast.WAFFLES.name())`

**Comment by swankjesse** [comment]
I'd prefer to offer a syntax or structure for arbitrary members (fields, enum constants, methods). That would also have the nice side effect of making static imports work.

**Comment by swankjesse** [comment]
https://github.com/square/javapoet/issues/317

**Comment by sormuras** [comment]
Hm, okay. But that `$E` does not exclude the full-fledged version. I think, they can easily live in parallel. Just you could only provide `$L` and not `$T`...

**Comment by swankjesse** [comment]
I think we should just ask callers to call `.name()` until we have a member reference model. I'm trying my best to keep the number of metacharacters in our fancy templating language down.

**Comment by sormuras** [comment]
Keeping the number of special tokens low is fine.

But you're the asking the caller to be verbose, as can be seen in the second comment. All information is present to infer the `$T.$L` pattern. And if I remember correctly, there are now and then some questions here and over at SO's regarding enum constant references.

**Comment by swankjesse** [comment]
Yeah, in this case I'm okay with requiring the caller to write more code in exchange for a smaller API. If/when we do member refs, we’ll grow the API in order to shrink the calling code.

**Comment by sormuras** [comment]
Alright, I'll file another PR with the `.name()` fixes, as the new `AnnotationSpec.get(Annotation)` is affected.
Poor`$E`. Will save the diff for future use. (-:
